src/Workspaces/Remote/ServiceHub/Host/RemotePasteTrackingService.cs

```diff
@@ -0,0 +1,105 @@
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Remote
+{
+    [Export(typeof(IPasteTrackingService))]
+    [Shared]
+    internal sealed class RemotePasteTrackingService : IPasteTrackingService
+    {
+        private readonly Dictionary<SourceTextContainer, (TextSpan? span, int referenceCount)> _trackedSpans = new Dictionary<SourceTextContainer, (TextSpan? span, int referenceCount)>();
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public RemotePasteTrackingService()
+        {
+        }
+
+        public bool TryGetPastedTextSpan(SourceTextContainer sourceTextContainer, out TextSpan textSpan)
+        {
+            lock (_trackedSpans)
+            {
+                var result = _trackedSpans.TryGetValue(sourceTextContainer, out var spanAndCount);
+                if (!result || spanAndCount.referenceCount == 0 || spanAndCount.span is null)
+                {
+                    textSpan = default;
+                    return false;
+                }
+
+                textSpan = spanAndCount.span.Value;
+                return result;
+            }
+        }
+
+        private void AddPastedTextSpanReference(SourceTextContainer container, TextSpan? pastedTextSpan)
+        {
+            lock (_trackedSpans)
+            {
+                if (!_trackedSpans.TryGetValue(container, out var spanAndCount))
+                {
+                    _trackedSpans[container] = (pastedTextSpan, 1);
+                    return;
+                }
+
+                if (spanAndCount.span != pastedTextSpan)
+                {
+                    // Cannot track two different pasted spans for the same container
+                    throw ExceptionUtilities.UnexpectedValue(pastedTextSpan);
+                }
+
+                Debug.Assert(spanAndCount.referenceCount > 0);
+                _trackedSpans[container] = (spanAndCount.span, spanAndCount.referenceCount + 1);
+            }
+        }
+
+        private void ReleasePastedTextSpanReference(SourceTextContainer container)
+        {
+            lock (_trackedSpans)
+            {
+                if (!_trackedSpans.TryGetValue(container, out var spanAndCount))
+                    throw ExceptionUtilities.UnexpectedValue(container);
+
+                Debug.Assert(spanAndCount.referenceCount > 0);
+                if (spanAndCount.referenceCount <= 1)
+                    _trackedSpans.Remove(container);
+                else
+                    _trackedSpans[container] = (spanAndCount.span, spanAndCount.referenceCount - 1);
+            }
+        }
+
+        internal Releaser SetPastedTextSpanForRemoteCall(SourceTextContainer container, TextSpan? pastedTextSpan)
+        {
+            AddPastedTextSpanReference(container, pastedTextSpan);
+            return new Releaser(this, container);
+        }
+
+        internal readonly struct Releaser : IDisposable
```

Comment:
nit: my preference is public on all nested symbols that are made 'internal' by virtue of their container.